Code/src/main/antlr4/nl/utwente/group10/haskell/typeparser/Type.g4

```diff
@@ -4,14 +4,20 @@ CT : [A-Z][a-z]+ ;
 VT : [a-z]+ ;
 WS : [ \t\r\n]+ -> skip ;
 
-type : functionType | compoundType ; // function type
-functionType : compoundType '->' type ;
+type : typeClasses ? innerType ;
+innerType : functionType | compoundType ; // function type
+functionType : compoundType '->' innerType ;
 compoundType : constantType | variableType | tupleType | listType | parenType ;
 
-tupleType : '(' type (',' type)+ ')' ; // tuple type, k>=2
-listType : '[' type ']' ;              // list type
-parenType : '(' type ')' ;             // type with parentheses
+tupleType : '(' innerType (',' innerType)+ ')' ; // tuple type, k>=2
+listType : '[' innerType ']' ;              // list type
+parenType : '(' innerType ')' ;             // type with parentheses
 
-constantType : typeConstructor (WS* type)* ;
+constantType : typeConstructor (WS* innerType)* ;
```

Comment:
That's true. On the other hand, it could give the impression that extra whitespace is *not* allowed elsewhere, which it is.